### PR TITLE
Please review

### PR DESCRIPTION
**Pull request by anoopezhap**
Please review

### CODE REVIEW

File: src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    public CustomersPage(WebDriver driver) {
+        super(driver);
+        this.driver = driver;
+        PageFactory.initElements(driver, this);
+    }
+
+    private String Time = "//span[text()=\"Time\"]";
+    private String ProjectInfo = "//span[normalize-space()='Project Info']";
+    private String Customers = "//a[text()=\"Customers\"]";
+    private String Add = "//button[normalize-space()='Add']";
```
Comment (thanexplore) [inline src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java]: Enhance variable names, for locators - more readable 
XpathAddCustomerButton

File: src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    private String NoOfRecords = "//div[@class=\"orangehrm-horizontal-padding orangehrm-vertical-padding\"]/span";
+    private String Name = "//div[@class=\"oxd-form-row\"]//input";
+    private String Description = "//div[@class=\"oxd-form-row\"]//textarea";
+    private String Save = "//button[@type=\"submit\"]";
+    private String Required = "//div[@class=\"oxd-form-row\"]//span";
+    private String YesDelete = "//div[@class=\"orangehrm-modal-footer\"]//button[2]";
+    private String SuccessMessage = "//div[@class=\"oxd-toast oxd-toast--success oxd-toast-container--toast\"]";
+    @FindBy(xpath = "//div[@class=\"oxd-table-card\"]")
```
Comment (thanexplore) [inline src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java:35]: extra one line indentation

File: src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        return isElementVisible(By.xpath(SuccessMessage));
+
+    }
+
+    public String mandatoryFields(String description) {
+        click(By.xpath(Add));
+        setText(By.xpath(Description), "Testinggggggg");
+        click(By.xpath(Save));
+        isElementVisible(By.xpath(Required));
+        return driver.findElement(By.xpath(Required)).getText();
```
Comment (thanexplore) [inline src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java]: creating a separate method for the same functionality in PageBase

File: src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    }
+
+    public int getNumberOfRecords() {
+        sleep(3000);
```
Comment (thanexplore) [inline src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java]: Used the similar iselement visible mechanism for the page elements,

File: src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        isElementVisible(By.xpath(NoOfRecords));
+        String Records = driver.findElement(By.xpath(NoOfRecords)).getText();
```
Comment (thanexplore) [inline src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java]: Separate common method can be created in pagebase

File: src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        String name = Records.split(Pattern.quote(")"))[0].split(Pattern.quote("("))[1].trim();
```
Comment (thanexplore) [inline src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java]: Separate method can be created in PageBase

File: src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        int NoOfRecords = Integer.parseInt(name);
+        System.out.println(NoOfRecords);
+        return NoOfRecords;
+    }
+
+    public int getActualNumberOfRecords() {
```
Comment (thanexplore) [inline src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java:81]: Generic method can be accomodated to PageBase

File: src/test/java/ca/qaguru/oranghrmbatch27/tests/CustomersTests.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+    //BATCH29-10
+    @Test
+    public void addNewCustomer()
+    {
+        LoginPage loginPage = new LoginPage(driver);
+        loginPage.login("Admin","admin123",true,null);
+        CustomersPage customersPage = new CustomersPage(driver);
+        customersPage.navigateToCustomer();
+        Boolean status = customersPage.addCustomer("Anoop Joseph","Flexible pouch manufacturer");
```
Comment (thanexplore) [inline src/test/java/ca/qaguru/oranghrmbatch27/tests/CustomersTests.java]: The data can be included in separate string variables

File: src/test/java/ca/qaguru/oranghrmbatch27/tests/CustomersTests.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+    //BATCH29-13
+    @Test
+    public void verifyNoOfRecords()
+    {
+        LoginPage loginPage = new LoginPage(driver);
+        loginPage.login("Admin","admin123",true,null);
+        CustomersPage customersPage = new CustomersPage(driver);
+        customersPage.navigateToCustomer();
+        Assert.assertEquals(customersPage.getNumberOfRecords(),customersPage.getActualNumberOfRecords());
```
Comment (thanexplore) [inline src/test/java/ca/qaguru/oranghrmbatch27/tests/CustomersTests.java:35]: Indentation fix

File: src/test/java/ca/qaguru/oranghrmbatch27/tests/CustomersTests.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    //BATCH29-14
+    @Test
+    public void verifyNoOfRecordsAfterAddingCustomer()
+    {
+        LoginPage loginPage = new LoginPage(driver);
+        loginPage.login("Admin","admin123",true,null);
+        CustomersPage customersPage = new CustomersPage(driver);
+        customersPage.navigateToCustomer();
+        int DisplayedNumber1 = customersPage.getNumberOfRecords();
+        customersPage.addCustomer("Deepu Joseph","Web developer");
```
Comment (thanexplore) [inline src/test/java/ca/qaguru/oranghrmbatch27/tests/CustomersTests.java]: To have the data in variables

**Review by thanexplore — COMMENTED** [review]
Kindly take care of review comments

File: src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java (hunk 1 above)
Comment (anoopezhap) [reply src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java]: Made changes to all the variables to make it more readable

File: src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java (hunk 2 above)
Comment (anoopezhap) [reply src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java:35]: Added an extra line for more clarity

File: src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java (hunk 3 above)
Comment (anoopezhap) [reply src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java]: getText method already exists in the pagebase. So I edited the code to call that method to get the text.

File: src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java (hunk 4 above)
Comment (anoopezhap) [reply src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java]: The code is working without the thread sleep

File: src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java (hunk 5 above)
Comment (anoopezhap) [reply src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java]: Fixed, by using getText() method fromt the page base

File: src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java (hunk 6 above)
Comment (anoopezhap) [reply src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java]: Created a reusable method inside the page base.

File: src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java (hunk 7 above)
Comment (anoopezhap) [reply src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java:81]: I am confused on what to do here. There is only one line of code.

File: src/test/java/ca/qaguru/oranghrmbatch27/tests/CustomersTests.java (hunk 8 above)
Comment (anoopezhap) [reply src/test/java/ca/qaguru/oranghrmbatch27/tests/CustomersTests.java]: Created separate  varibles to store the information

File: src/test/java/ca/qaguru/oranghrmbatch27/tests/CustomersTests.java (hunk 9 above)
Comment (anoopezhap) [reply src/test/java/ca/qaguru/oranghrmbatch27/tests/CustomersTests.java:35]: It is fixed.

File: src/test/java/ca/qaguru/oranghrmbatch27/tests/CustomersTests.java (hunk 10 above)
Comment (anoopezhap) [reply src/test/java/ca/qaguru/oranghrmbatch27/tests/CustomersTests.java]: Created separate varibles to store the information

File: src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    public boolean addCustomer(String name, String description) {
+        click(By.xpath(xpathAddButton));
+        setText(By.xpath(xpathNameCell), name);
+        setText(By.xpath(xpathDescriptionCell), description);
+        click(By.xpath(xpathSaveButton));
+        return isElementVisible(By.xpath(xpathSuccessMessage));
+
+    }
+
+    public String mandatoryFields(String description) {
```
Comment (merinj20) [inline src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java:66]: String description is not used in the method.

**Review by merinj20 — COMMENTED** [review]
review point added

File: src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java (hunk 11 above)
Comment (anoopezhap) [reply src/test/java/ca/qaguru/oranghrmbatch27/pages/CustomersPage.java:66]: Fixed it

**Review by thanexplore — APPROVED** [review]

**Pull request merged**